src/runtime_src/core/include/experimental/xrt_xclbin.h

```diff
@@ -19,11 +19,199 @@
 #define _XRT_XCLBIN_H_
 
 #include "xrt.h"
+#include "experimental/xrt_uuid.h"
 
 #ifdef __cplusplus
+# include <memory>
+# include <vector>
+# include <string>
+#endif
+
+/**
+ * typedef xrtXclbinHandle - opaque xclbin handle
+ */
+typedef void* xrtXclbinHandle;
+
+#ifdef __cplusplus
+namespace xrt {
+
+class xclbin_impl;
+class xclbin
+{
+public:
+  
+  /**
+   * xclbin() - Copy ctor
+   */
+  xclbin(const xclbin& rhs)
+    : handle(rhs.handle)
+  {}
+  
+  /**
+   * xclbin() - Move ctor
+   */
+  xclbin(xclbin&& rhs)
+    : handle(std::move(rhs.handle))
+  {}
+  
+  /**
+   * operator= () - Move assignment
+   */
+  xclbin&
+  operator=(xclbin&& rhs)
+  {
+    handle = std::move(rhs.handle);
+    return *this;
+  }
+
+  /**
+   * xclbin() - Constructor from an xclbin filename
+   *
+   * @filename:  path to the xclbin file
+   *
+   * The xclbin file must be accessible by the application. An exception is thrown file not found
+   */
+  explicit xclbin(const std::string& filename);
+  
+  /**
+   * xclbin() - Constructor from raw data
+   *
+   * @data: raw data of xclbin
+   *
+   * The raw data of the xclbin can be deleted after calling the constructor.
+   *
+   */
+  xclbin(const std::vector<char>& data);
+
+  /**
+   * get_cu_names() - Get CU names of xclbin
+   *
+   * Return: A list of CU Names in order of increasing base address.
+   *
+   * An exception is thrown if the data is missing.
+   */
+  const std::vector<std::string>
+  get_cu_names() const;
+
+  /**
+   * get_xsa_name() - Get Xilinx Support Archive (XSA) Name of xclbin
+   *
+   * Return: Name of XSA
+   *
+   * An exception is thrown if the data is missing.
+   */
+  const std::string
+  get_xsa_name() const;
+
+  /**
+   * get_uuid() - Get the uuid of the xclbin
+   *
+   * Return: UUID of xclbin
+   *
+   * An exception is thrown if the data is missing.
+   */
+  uuid
+  get_uuid() const;
+
+  /**
+   * get_data() - Get the raw data of the xclbin
+   *
+   * Return: The raw data of the xclbin
+   *
+   * An exception is thrown if the data is missing.
+   */
+  const std::vector<char>&
+  get_data() const;
+
+public:
+  xclbin() = delete;
```

Comment:
You may want to allow an empty xrt::xclbin object that can be assigned to later.
Provide operator bool() for testing if the object has content.

Reply:
Curious... what is the semantics of an `xclbin` then?

Reply:
@stsoe done in local sandbox, will push in next PR.